upup/pkg/fi/cloudup/gce/tpm/gcetpmsigner/tpmauthenticator.go

```diff
@@ -0,0 +1,150 @@
+	"encoding/pem"
+	"fmt"
+	"time"
+
+	"cloud.google.com/go/compute/metadata"
+	"github.com/google/go-tpm-tools/client"
+	"k8s.io/klog/v2"
+	"k8s.io/kops/pkg/apis/nodeup"
+	"k8s.io/kops/pkg/bootstrap"
+)
+
+type tpmAuthenticator struct {
+	projectID string
+	zone      string
+	instance  string
+}
+
+var _ bootstrap.Authenticator = &tpmAuthenticator{}
+
+func NewTPMAuthenticator() (bootstrap.Authenticator, error) {
+	projectID, err := metadata.ProjectID()
+	if err != nil {
+		return nil, fmt.Errorf("error getting projectID from metadata: %w", err)
+	}
+	zone, err := metadata.Zone()
+	if err != nil {
+		return nil, fmt.Errorf("error getting zone from metadata: %w", err)
+	}
+	instance, err := metadata.InstanceName()
+	if err != nil {
+		return nil, fmt.Errorf("error getting instance from metadata: %w", err)
+	}
+
+	return &tpmAuthenticator{
+		projectID: projectID,
+		zone:      zone,
+		instance:  instance,
+	}, nil
+}
+
+func (a *tpmAuthenticator) SignBootstrapRequest(req *nodeup.BootstrapRequest) (*bootstrap.SignedRequestData, error) {
+	tpmStart := time.Now()
+
+	tpmDevice, err := openTPM()
+	if err != nil {
+		return nil, fmt.Errorf("failed to open TPM: %w", err)
+	}
+	defer tpmDevice.Close()
+
+	key, err := client.GceAttestationKeyRSA(tpmDevice)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get GCE RSA attestation key from TPM: %w", err)
+	}
+	defer key.Close()
+
+	klog.V(2).Infof("attestation key is %v", debugToPEM(key.PublicKey()))
+
+	klog.Infof("TPM initialization took %v", time.Since(tpmStart))
+
+	if req.Claim == nil {
+		req.Claim = &nodeup.Claim{}
+	}
+
+	{
+		node := &nodeup.NodeClaim{
+			GCPProjectID: a.projectID,
+			Zone:         a.zone,
+			Instance:     a.instance,
+		}
+
+		payload, err := json.Marshal(node)
+		if err != nil {
+			return nil, fmt.Errorf("failed to marshal node claim: %w", err)
+		}
+		signature, err := tpmSign(key, payload)
+		if err != nil {
+			return nil, fmt.Errorf("failed to sign node claim: %w", err)
+		}
+		req.Claim.Node = node
+		req.Claim.NodeSignature = signature
+	}
+
+	request := &bootstrap.SignedRequestData{}
+	b, err := json.Marshal(req)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal request: %w", err)
+	}
+	request.Body = b
+	return request, nil
+}
+
+// tpmSign performs a TPM signature with the tpmKey, and sanity checks the result.
+func tpmSign(tpmKey *client.Key, payload []byte) ([]byte, error) {
+	beforeSign := time.Now()
+	signature, err := tpmKey.SignData(payload)
+	if err != nil {
+		return nil, fmt.Errorf("failed to sign data with TPM: %w", err)
+	}
+
+	// double-check the signature.
+	keyRSA, ok := tpmKey.PublicKey().(*rsa.PublicKey)
+	if !ok {
+		return nil, fmt.Errorf("unexpected type for attestation key, got %T, want rsa.PublicKey", tpmKey.PublicKey())
+	}
+
+	attestHash := sha256.Sum256(payload)
+	if err := rsa.VerifyPKCS1v15(keyRSA, crypto.SHA256, attestHash[:], signature); err != nil {
+		return nil, fmt.Errorf("signature double-check failed: %w", err)
```

Comment:
What's the purpose of the double-check?

Reply:
Removed, it was in the gce provider but I agree it wasn't doing anything.